Multiple drivers

EtoMart opened the pull request:
done task

EtoMart committed:
Add custom-validation class form dates error
change form.component.ts validation & drivers generation

EtoMart committed:
done task

EtoMart committed:
clean imports, quots
add route after successful validate

marden007 commented on src/app/main/classes/form-control-value-accessor-adapter.ts:
```diff
@@ -7,11 +7,11 @@ import {
+  }
 
   writeValue(val: any): void {
     if (val) {
```
``` public  writeValue(...```

marden007 replied:
+  для того чтобы уменьшить вложенность при проверке условия в начале метода принято использовать логическое отрицание
```
if (!val) {
return;
}
```

marden007 commented on src/app/main/main.module.ts:
```diff
@@ -1,10 +1,11 @@
 
 @NgModule({
   declarations: [],
   imports: [CommonModule, ReactiveFormsModule, RouterModule],
```
каждый модуль в новой строке

marden007 commented on src/app/main/modules/driver-module/component/driver-component.component.html:
```diff
@@ -1,44 +1,48 @@
   <fieldset>
     <label for="isInsured">
-      <input formControlName="isInsured" id="isInsured" type="checkbox"/> является страхователем
+      <input formControlName="isInsured" id="isInsured" type="checkbox" /> является страхователем
```
```
<input formControlName="isInsured" 
       type="checkbox"
       id="isInsured"/>
       является страхователем
</label>
```

marden007 replied:
такое же правило как и для модулей/компонентов - для отслеживания изменений каждый атрибут на новой строке

marden007 commented on src/app/main/modules/driver-module/component/driver-component.component.html:
```diff
@@ -1,44 +1,48 @@
         <label for="startExpDate">Дата начала стажа</label>
-        <input formControlName="startExpDate" id="startExpDate" type="date"/>
+        <input
+          [ngClass]="{'custom-invalid': form.errors?.dates === true && form.controls['startExpDate'].touched && form.controls['birthday'].touched && form.controls['startExpDate'].dirty}"
```
`form.errors?.dates === true && form....` можно сократить до `form.errors?.dates && form.....`

marden007 commented on src/app/main/modules/driver-module/component/driver-component.component.ts:
```diff
@@ -72,5 +59,6 @@ export class DriverComponent
   }
 
-  ngOnDestroy(): void {}
+  ngOnDestroy(): void {
```
неиспользуемый метод - удалить

marden007 commented on src/app/main/modules/driver-module/component/driver-component.component.ts:
```diff
@@ -41,16 +29,15 @@ export class DriverComponent
       isInsured: [false, []],
     },
-    { validators: this.startExpDateValidator('startExpDate', 'birthday') }
+    { validators: this.startExpDateValidator('startExpDate', 'birthday') },
```
метод валидации можно вынести из компонента, например рядом с самим компонентом
сейчас у тебя в ё`driver-module` каталог `component` для одного компонента, но у модуля может быть много компонентов
1) переименовать каталог `component` в `components`
2) в каталоге создать еще один каталог - `driver` и запихнуть туда `driver-component`
3) создать в этом каталоге(`driver`) каталог `validators` и создать в нем файл `start-exp-date.validator.ts`
4) перенести метод `startExpDateValidator` в новый файл

в результате если мы в будущем захотим добавить еще один компонент или еще один валидатор у нас уже будет где это сделать

marden007 commented on src/app/main/modules/form/component/form.component.html:
```diff
@@ -1,23 +1,20 @@
-    <app-driver-component *ngIf="driversCount > 2" formControlName='driver3'></app-driver-component>
-    <app-driver-component *ngIf="driversCount > 3" formControlName='driver4'></app-driver-component>
-    <app-driver-component *ngIf="driversCount > 4" formControlName='driver5'></app-driver-component>
+  <div *ngFor="let driver of driversCount">
```
:like

marden007 commented on src/app/main/modules/form/component/form.component.ts:
```diff
@@ -1,83 +1,117 @@
+    this.initialDriverChange();
+  }
+
+  ngAfterContentInit(): void {
```
мусор - удалить

EtoMart committed:
deleted formControlValueAccessorAdapter class
relocate driverComponent
separated custom driver validator
corrected imports
added line breaks by attributes to html
deleted unused methods

marden007 commented on src/app/main/modules/driver-module/components/driver/driver-component.component.html:
```diff
@@ -0,0 +1,146 @@
+<ng-container [formGroup]='form'
+              class='container'>
+  <fieldset>
+    <label for='isInsured'>
```
в html двойные кавычки)

EtoMart replied:
Done

marden007 commented on src/app/main/modules/driver-module/components/driver/driver-component.component.html:
```diff
@@ -0,0 +1,146 @@
+    </div>
+  </fieldset>
+  <div
+    *ngIf="
```
чуть-чуть перевыполнил) 
первый атрибут остается вместе с тегом + закрывающую скобку не отрываем (>)
```
<div *ngIf="form.controls['firstName'].invalid && form.controls['firstName'].touched"
  class="alert alert-danger">
....
</div>
```
если условие большое(строка больше 140 символов), то можно само условие разнести по строчкам
```
  <div *ngIf="form.errors?.dates && form.controls['startExpDate'].touched &&
      form.controls['birthday'].touched && form.controls['startExpDate'].dirty"
    class='alert alert-danger'>
</div>
```

EtoMart replied:
косяк, тут явно намудрил с настройками стилей :)

marden007 commented on src/app/main/modules/driver-module/components/driver/driver-component.component.ts:
```diff
@@ -0,0 +1,41 @@
+  styleUrls: ['./driver-component.component.scss'],
+})
+export class DriverComponent implements OnInit {
+  form: FormGroup = this.formBuilder.group(
```
`public form: FormGroup ...`

EtoMart replied:
done

marden007 commented on src/app/main/modules/driver-module/components/driver/driver-component.component.ts:
```diff
@@ -0,0 +1,41 @@
+  constructor(private formBuilder: FormBuilder) {
+  }
+
+  ngOnInit(): void {
```
`public ngOnInit(....`

EtoMart replied:
done

marden007 commented on src/app/main/modules/driver-module/components/driver/driver-component.component.ts:
```diff
@@ -0,0 +1,41 @@
+
+  ngOnInit(): void {
+    setTimeout(() => {
+      this.form.updateValueAndValidity();
```
хак, можно без него обойтись

EtoMart replied:
хм, и правда :)

marden007 commented on src/app/main/modules/form/component/form.component.ts:
```diff
@@ -1,83 +1,100 @@
+
+  private getDrivers(): void {
+    const components = this.driverComponents.toArray();
+    for (let i = 0; i < components.length; i++) {
```
для переборки массивов в Js принято использовать forEach(если нужно для каждого элемента массива вызвать какой-нибудь метод), map(если нужно преобразовать текущий массив и получить на его основе другой - пример - из массива водителей получить массив id водителей для сохранения на бек) и др.
Это избавляет от хлопотов с этими индексами (i, j,k) их нужно инициализировать, определять условие.
+Можно запутаться во вложенных переборках массива.
Не бойся именовать переменные, методы большими словосочетаниями, если без этого теряется их содержимое(tempArray  и ему подобное нужно исключить, т.к. если ты через какое-то время откроешь код ты не вспомнишь что имел в виду)

пример:
```
private getDrivers(): void {
    const driversComponents = this.driverComponents.toArray();
    driversComponents.forEach((component, index) =>  {
       component.form.setValue(this.drivers[index]);
       this.changeDetectorRef.detectChanges();
    }
}
```
или даже можно избавиться от переменной, которую мы инициализируем и используем только один раз на следующей строчке
```
private getDrivers(): void {
    this.driverComponents.toArray().forEach((component, index) =>  {
       component.form.setValue(this.drivers[index]);
       this.changeDetectorRef.detectChanges();
    }
}
```

EtoMart replied:
Прикольно ). С фор ич знаком, но редко использую )

marden007 commented on src/app/main/modules/form/component/form.component.ts:
```diff
@@ -1,83 +1,100 @@
+{
+  @ViewChildren('driverComponent') driverComponents: QueryList<DriverComponent>;
+  drivers = [];
+  driversCount = [0];
```
название переменной `driversCount` не совпадает с действительностью и подразумевает количество  водятлов
```
public driversCount: number = 1;
```

в такой реализации можно вообще от этой переменной отказаться и использовать массив drivers

EtoMart replied:
done

marden007 commented on src/app/main/modules/form/component/form.component.ts:
```diff
@@ -1,83 +1,100 @@
+    }
   }
 
   ngOnInit(): void {
```
в классах идут сначала публичные методы, потом приватные
вот примерно как должны располагаться переменные, методы и т.д

```
@Component({...})
export class DriverComponent {
   @Input()
    public isOwner: boolean = false;

   @Output()
    public onDriverChanged: EventEmitter<void> = new EventEmitter();

    public drivers: Driver [] = [];
    private isSaved: boolean = false;

    constructor(private changeDetectorRef: ChangeDetectorRef,) {
    }

   public ngOnInit(): void {...}

   public ngAfterViewInit(): void {...}

   public saveDriver(): void {...}

   private init(): void {....}
}
```

EtoMart replied:
Учту, не знал :)

marden007 reviewed (changes requested):
ну во, с каждой итерацией лучше и лучше
причесать немного, поправить отображение водителей и аппрув)

EtoMart committed:
replaced for with forEach
added methods
fixed html

EtoMart committed:
Change DriversCount

EtoMart committed:
rename driverComponentsCount to driverComponents

EtoMart committed:
add sex directive

EtoMart committed:
Done driver SexDirective

EtoMart committed:
correct driver sex directive

EtoMart committed:
add httpDriverService

EtoMart committed:
add HttpDriverData interface
add httpGet method

EtoMart committed:
add check flag method to httpDriver
add unsubscribe to subscriptions
make GetDrivers button disable by default
add OnDestory to FormComponent for unsubscribe
Clear from console.log

marden007 commented on src/app/main/directives/select-sex.directive.ts:
```diff
@@ -0,0 +1,23 @@
+
+  constructor(private el: ElementRef) { }
+
+  @Output() childEvent = new EventEmitter();
```
непонятный output: название не содержит инфы по какому случаю он срабатывает  + не указан возвращаемый тип
примерно так можно обозвать:
```
@Output
public onGenderDetect: EventEmitter<string> = new EventEmitter<string>();
```

marden007 commented on src/app/main/interfaces/http-driver-data.ts:
```diff
@@ -0,0 +1,31 @@
+export interface HttpDriverData {
+  id?: any;
```
прям все поля необязательные?)

EtoMart replied:
тут сложности ...

marden007 commented on src/app/main/modules/driver-module/driver-module.module.ts:
```diff
@@ -1,12 +1,14 @@
-  imports: [CommonModule, FormsModule, ReactiveFormsModule],
+  imports: [CommonModule,
+    FormsModule,
+    ReactiveFormsModule, MainModule],
```
:(

marden007 replied:
импорты поехали

marden007 commented on src/app/main/services/form.service.ts:
```diff
@@ -18,6 +18,7 @@ export class FormService {
       middleName: '',
       oldDriverLicence: false,
       startExpDate: '',
+      sex: 'male',
```
для пола лучше завести Enum т.к.: 
1) дублирование строк "male" может привести к опечатке
2) при изменении в  api строки 'male' на "M" потребует больше времени
3) в наше время может добавиться еще какой-нибудь пол))

EtoMart replied:
1) Окей
2) Окей
3) Осуждаю :)

marden007 commented on src/app/main/services/http-driver.service.ts:
```diff
@@ -0,0 +1,69 @@
+        Authorization: driverHttpAuthToken.token,
+      }),
+    };
+    return this.http
```
норм :like

marden007 commented on src/app/main/services/http-driver.service.ts:
```diff
@@ -0,0 +1,69 @@
+      );
+  }
+
+  public getData(id: string): any {
```
тут надо бы возвращать не any, а DriverDataInterface, для того чтобы подставлять данные в форму

EtoMart replied:
а косяк (

marden007 commented on src/app/main/main.module.ts:
```diff
@@ -1,10 +1,20 @@
+  declarations: [
+    SelectSexDirective,
+  ],
+  imports: [CommonModule,
```
тут тоже импорты с новой строки
```
imports: [
    CommonModule,
    ReactiveFormsModule,
    RouterModule,
    HttpClientModule,
],

EtoMart committed:
Change getData method - add map to return DriverDataInterface

EtoMart committed:
correct imports tabs
